koa/koa.d.ts

```diff
@@ -1,182 +1,654 @@
-            status: number;
-            type: string;
-            writable: boolean;
+        interface Context extends BaseContext, Koa.Context {
+            app: Application;
+            request: Request;
+            response: Response;
+            req: http.IncomingMessage;
+            res: http.ServerResponse;
+            originalUrl: string;
+            cookies: cookies.ICookies;
+            accept: any;
+            state: any;
+        }
 
-            charset?: string;
-            length?: number;
+        interface Application extends EventEmitter, Koa.Application {
+            proxy: boolean;
+            middleware: Middleware[];
+            subdomainOffset: number;
+            env: string;
+            context: BaseContext;
+            request: BaseRequest;
+            response: BaseResponse;
+            silent: boolean;
+            keys: string[];
 
-            append(field: string, val: string | string[]): void;
-            attachment(filename?: string): void;
-            get(field: string): string;
-            is(): string[];
-            is(arg: string): void | string;
-            is(arg: string[]): void | string;
-            is(...args: string[]): void | string;
-            redirect(url: string, alt?: string): void;
-            remove(field: string): void;
-            set(field: string, val: string | string[]): void;
-            set(field: any): void;
-            vary(field: string): void;
+            /**
+             * Shorthand for:
+             *
+             *    http.createServer(app.callback()).listen(...)
+             */
+            listen(port: number, hostname?: string, backlog?: number, callback?: Function): http.Server;
+            listen(port: number, hostname?: string, callback?: Function): http.Server;
+            listen(path: string, callback?: Function): http.Server;
+            listen(handle: any, listeningListener?: Function): http.Server;
+            listen(): http.Server;
 
-            toJSON(): any;
+            /**
+             * Return JSON representation.
+             * We only bother showing settings.
+             */
             inspect(): any;
+
+            /**
+             * Return JSON representation.
+             * We only bother showing settings.
+             */
+            toJSON(): any;
+
+            /**
+             * Use the given middleware `fn`.
+             *
+             * Old-style middleware will be converted.
+             */
+            use(middleware: Middleware): Application;
+
+            /**
+             * Return a request handler callback
+             * for node's native http server.
+             */
+            callback(): (req: http.IncomingMessage, res: http.ServerResponse) => void;
+
+            /**
+             * Default error handler.
+             */
+            onerror(err: Error): void;
+        }
+
+        interface Middleware {
+            (context: Context, next?: () => Promise<void>): any;
```

Comment:
@jkeylu Could you please make `next` non-optional? The way koa implements middleware can be found [here](https://github.com/koajs/compose/blob/02c4d664c4a0f07c8b96d0010cd0f1c465190de7/index.js#L43-L45) and as you can see `next` is always provided.